sled-storage/src/manager.rs

```diff
@@ -590,9 +663,145 @@ impl StorageManager {
             info!(self.log, "KeyManager ready, but no ledger detected");
         }
 
+        // We don't load any configuration for datasets, since we aren't
+        // currently storing any dataset information in-memory.
+        //
+        // If we ever wanted to do so, however, we could load that information
+        // here.
+
         Ok(())
     }
 
+    async fn datasets_ensure(
+        &mut self,
+        mut config: DatasetsConfig,
+    ) -> Result<DatasetsManagementResult, Error> {
+        let log = self.log.new(o!("request" => "datasets_ensure"));
+
+        // Ensure that the datasets arrive in a consistent order
+        config.datasets.sort_by(|a, b| a.id.partial_cmp(&b.id).unwrap());
+
+        // We rely on the schema being stable across reboots -- observe
+        // "test_datasets_schema" below for that property guarantee.
+        let ledger_paths = self.all_omicron_dataset_ledgers().await;
+        let maybe_ledger =
+            Ledger::<DatasetsConfig>::new(&log, ledger_paths.clone()).await;
+
+        let mut ledger = match maybe_ledger {
+            Some(ledger) => {
+                info!(
+                    log,
+                    "Comparing 'requested datasets' to ledger on internal storage"
+                );
+                let ledger_data = ledger.data();
+                if config.generation < ledger_data.generation {
+                    warn!(
+                        log,
+                        "Request looks out-of-date compared to prior request"
+                    );
```

Comment:
Might add the two generations to this log message.

Reply:
Done